ext/spl/spl_observer.stub.php

```diff
@@ -4,155 +4,155 @@
 
 interface SplObserver
 {
-    /** @return void */
-    public function update(SplSubject $subject);
+    /** @tentative-return-type */
+    public function update(SplSubject $subject): void;
 }
 
 interface SplSubject
 {
-    /** @return void */
-    public function attach(SplObserver $observer);
+    /** @tentative-return-type */
+    public function attach(SplObserver $observer): void;
 
-    /** @return void */
-    public function detach(SplObserver $observer);
+    /** @tentative-return-type */
+    public function detach(SplObserver $observer): void;
 
-    /** @return void */
-    public function notify();
+    /** @tentative-return-type */
+    public function notify(): void;
 }
 
 class SplObjectStorage implements Countable, Iterator, Serializable, ArrayAccess
 {
-    /** @return void */
-    public function attach(object $object, mixed $info = null) {}
+    /** @tentative-return-type */
+    public function attach(object $object, mixed $info = null): void {}
 
-    /** @return void */
-    public function detach(object $object) {}
+    /** @tentative-return-type */
+    public function detach(object $object): void {}
 
-    /** @return bool */
-    public function contains(object $object) {}
+    /** @tentative-return-type */
+    public function contains(object $object): bool {}
 
-    /** @return int */
-    public function addAll(SplObjectStorage $storage) {}
+    /** @tentative-return-type */
+    public function addAll(SplObjectStorage $storage): int {}
 
-    /** @return int */
-    public function removeAll(SplObjectStorage $storage) {}
+    /** @tentative-return-type */
+    public function removeAll(SplObjectStorage $storage): int {}
 
-    /** @return int */
-    public function removeAllExcept(SplObjectStorage $storage) {}
+    /** @tentative-return-type */
+    public function removeAllExcept(SplObjectStorage $storage): int {}
 
-    /** @return mixed */
-    public function getInfo() {}
+    /** @tentative-return-type */
+    public function getInfo(): mixed {}
 
-    /** @return void */
-    public function setInfo(mixed $info) {}
+    /** @tentative-return-type */
+    public function setInfo(mixed $info): void {}
 
-    /** @return int */
-    public function count(int $mode = COUNT_NORMAL) {}
+    /** @tentative-return-type */
+    public function count(int $mode = COUNT_NORMAL): int {}
 
-    /** @return void */
-    public function rewind() {}
+    /** @tentative-return-type */
+    public function rewind(): void {}
 
-    /** @return bool */
-    public function valid() {}
+    /** @tentative-return-type */
+    public function valid(): bool {}
 
-    /** @return int */
-    public function key() {}
+    /** @tentative-return-type */
+    public function key(): int {}
```

Comment:
I can easily refactor this to use composition instead of inheritance https://github.com/phan/phan/blob/v5/src/Phan/Library/Map.php#L30-L40 ( at the cost of a bit of speed) and suppress for now via annotation

But I wonder how many things override SplObjectStorage right now to change key/current to something else in the absense of a native `Map` type for objects outside of ds/its polyfill.

Aside: Would it make sense to add a new method SplObjectStorage->currentValue() to avoid the need to call $splObjectStorage->offsetGet($splObjectStorage->current())

Reply:
While I hate the interface that SplObjectStorage exposes (it probably is that way because it predates the non-int/string iterator key support I added in PHP 5.5 or so), your class still clearly violates the API -- if something accepts an SplObjectStorage and is passed your Map instead, it will not behave correctly. So I do think that switching the implementation to use composition would be the right move here.

Reply:
Never mind, It turns out the functionality I was thinking of already exists as https://www.php.net/manual/en/splobjectstorage.getinfo.php